### ISSUE TITLE
Platform: MinGW (aka Windows 32 Support)

### ISSUE THREAD

**Issue by bakercp** [olab-io/ofSketch#49]
We currently need to figure out how to build on windows using makefiles.  This should be possible, as I believe codeblocks, etc does it -- but this needs more research.

**Comment by AmnonOwed** [comment]
:+1: +1 on this! :)

Any indication on when this feature may be implemented?

_EDIT: Just noticed the milestone of September 1, 2014. Is that the eta?_

**Comment by bakercp** [comment]
Yes, September is my goal. Sooner would be great. If you have experience with makefiles mingw / msys and windows we love some help :)

**Comment by AmnonOwed** [comment]
Thanks for the quick reply. Unfortunately I have zero experience with that, so I can't be of much help here. Just started really researching openFrameworks and this project seems like a good step from Processing to oF. Good luck!

**Comment by bakercp** [comment]
This will indeed be an easy jump from our Processing!  Thanks for your interest!

**Comment by mikewesthad** [comment]
Here's the format for building a project using CLI with CodeBlocks on windows:

> pathToCodeBlockExe pathToWorkspaceFile --build

[Reference for CLI for codeblocks](http://wiki.codeblocks.org/index.php?title=Code::Blocks_command_line_arguments)

**Comment by mikewesthad** [comment]
Re: discussion on using MSYS and MinGW to make.  I started with a clean install of MinGW, MYSTS, openFrameworks repo and ofSketch repo.
1. Used these settings for the installation of MinGW and MSYS:

![screenshot 2014-08-03 10 47 09](https://cloud.githubusercontent.com/assets/1131802/3791132/3aa5f43c-1b2a-11e4-8884-1d450087ea7a.png)
1. Added C:\MinGW\bin to PATH user variable
2. Used @bakercp's config.shared.mk and config.win_cb.default.mk.  (With the modification of one line in config.shared.mk from "MINGW_NT-6.2" to "MINGW_NT-6.1". 
3. Using MSYS, I was able to make openFrameworks without any errors.  (No need for the "Additions for Code::Blocks to work with openFrameworks" from the openFrameworks for Code::Blocks installation page)

Not sure how to test that the building worked, other than to try making ofSketch.  I'm downloading the necessary addons and trying that now.

**Comment by mikewesthad** [comment]
Oh well.  Hit a wall, and it's over my head.

Making ofSketch errored, but I'm not sure why.  A bunch of warnings were thrown, plus one error log form JSONRPC.  
![screenshot 2014-08-03 11 40 11](https://cloud.githubusercontent.com/assets/1131802/3791198/39e74970-1b2e-11e4-8de8-32a9eda8b8a7.png)

I tried borrowing the Makefile and config.make files from ofSketch and using them to build a simple project that doesn't use addons, but that also errored out.  It got further along than ofSketch (it gets to the point where it is compiling for release), but then it threw a whole bunch of undefined references plus a bad reloc address error.  It fails at linking for win_cb.  Edit: adding the Code::Block additional files didn't fix it. 
![screenshot 2014-08-03 11 46 51](https://cloud.githubusercontent.com/assets/1131802/3791201/b024b1b8-1b2e-11e4-9180-f8daac26c52e.png)

**Comment by bakercp** [comment]
Hey @mikewesthad thanks for the work on this.  We're pretty much in the same place at this point.  

Just for reference, I am working on this from my "win-fix" oF branch (https://github.com/bakercp/openFrameworks/blob/win-fix).

So the following steps need to be investigated (let me know if you want to tackle some):
- [ ] Contact @arturoc and find out what version of mingw (and compiler flags) were used to compile all of the win_cb libs (OR) recompile all of them anyway using the apothecary formulas (which may or may not yet exist for all of the libs).  Ideally, this mingw produced binaries _should_ work with win_cb.  I don't want to create a whole new platform ...
- [ ] Figure out how to to use the custom project make file (https://github.com/bakercp/openFrameworks/blob/win-fix/libs/openFrameworksCompiled/project/win_cb/config.win_cb.default.mk) to link into the required libs -- I'm assuming this is similar to other platforms -- I just haven't had a moment to find the static libs and add them.  I'll be taking my hints from the other platform specific project files.
- [ ] Get rid of the OSX specific stuff in https://github.com/bakercp/openFrameworks/blob/win-fix/libs/openFrameworksCompiled/project/win_cb/config.win_cb.default.mk.

On the bike ride home yesterday after the hack-a-thon I thought more about using the command line codeblocks / VS approach and there are a lot of shortcomings that will make it worth it to keep pushing on this.  Writing a whole separate error parsing system is one of the problems.  Also, in the future we'd like to be using clang / llvm to compile all of this stuff, which will (theoretically) be easier if we have a working minGW makefile system.  Finally, having a working mingw makefile workflow would benefit the larger community ...

**Comment by mikewesthad** [comment]
I'd love to jump in with these, but my understanding of all of this is pretty basic at the moment, so I'm not sure where I can be helpful. 

I did find this tool, [cbp2make](http://sourceforge.net/projects/cbp2make/), that I thought might be helpful in figuring out what the current config.win_cb.default.mk is missing.  Here's an example of what the app spits out: [pastebin](http://pastebin.com/0XSyTnpG).  I tried putting all the libs from there into the win mk.  I ended up with a shorter (though different) list of linker errors, no clue if it's better or moving in the right direction.

**Comment by bakercp** [comment]
@mikewesthad Cool!  That makefile is actually super helpful.  I'll spend some time w/ it and see what I can learn.

**Comment by arturoc** [comment]
> Contact @arturoc and find out what version of mingw (and compiler flags) were used to compile all of the win_cb libs

the way i've been compiling libs in windows is by using msys but with the version of mingw that come with codeblocks, the latest ones with codeblocks 12.11. i just delete the mingw folder inside msys and substitute it with a copy of the one that comes with codeblocks. that way it's easy to use cmake, autotools, makefiles...

the reason for compiling with the version of mingw that comes with codeblocks is that some libraries, mostly the ones that use the std won't be compatible if you compile them with the mingw that comes with msys

**Comment by bakercp** [comment]
@arturoc Thanks Arturo -- that's what I suspected.  I'll give that a try!

**Comment by mikewesthad** [comment]
I'm still erroring out with the same linking errors using the Code::Blocks 12.11 MinGW.  Based on your second checkbox, the win make file needs to have some additional parameters that point to the static core libraries?

For documentation's sake, here's some version info for the MinGW from Code::Blocks 12.11 (all of which are older versions than the ones in the MinGW standalone install):

```
gcc version 4.7.1 (tdm-1)
GNU gdb (GDB) 7.5
GNU ld (GNU Binutils) 2.22
GNU Make 3.82
```

**Comment by bakercp** [comment]
More clues here -- 
- MSYS2 seems to be an interesting way forward, especially for compiling using apothecary.  It is actively maintained and it has a package manager (pacman), etc.  I installed it, but haven't tried any of this yet.  http://sourceforge.net/projects/msys2/.  I'm not sure of the best way to get MSYS2 to use the codeblocks 12.11 MinGW executables -- probably a matter of setting up paths, but I'm not going to be able to try it right away.
- Poco 1.5+ has a much more sophisticated process object, allowing us to set environmental variables, working directories, etc.  This should be helpful on windows.  See here http://pocoproject.org/docs/Poco.Process.html#9897

**Comment by mikewesthad** [comment]
Okay, a lot of stumbling, but here's some progress.  I can get an openFrameworks project to compile using MSYS with the CodeBlock 12.11 MinGW.  Didn't test it with any other versions.

Using the build commands from Code::Blocks as a guide, I got a sample project to compile.    It builds and runs (and spits out the same warnings as Code::Blocks).  See [repo](https://github.com/mikewesthad/openFrameworksWinMakeExample).  If you throw that inside of /apps/myapps/, it should work (so long as you compiled openFrameworks previously).  The Makefile doesn't rebuild openFrameworks or hook into the openFrameworks Makefiles.  I also couldn't get xcopy flags working when copying the DLLs over to the bin, but oh well, that's easy to sort out later.

**Comment by bakercp** [comment]
@mikewesthad Wow, great news!  Did you end up using MSYS or MSYS2?

**Comment by mikewesthad** [comment]
@bakercp I used MSYS.  

As proof-of-concept that the whole oF make system should work with MSYS+MinGW, I cleared openFrameworks/lib/win_cb/ and called the oF Makefile.  It crapped out at the linker as expected. But then I re-called the linking stage with my own command.  It builds a working exe.  I take that to mean that we can add/tweak commands on the oF Makefile system, and volia.

**Comment by bakercp** [comment]
Very nice.  Sounds like we need to just tweak.  I am taking a few days off
from ofSketch to focus on other projects, but I'll be back on this next
week ... keep me posted if you make any more progress or have any specific
questions about our magical makefile system :)

---

http://christopherbaker.net

**Comment by mikewesthad** [comment]
Will do.  It's going to take me a while to digest the magic of the makefile system...

**Comment by mikewesthad** [comment]
**tl;dr:** 
Why would the order of libraries break linking on windows, but not on any other system?  What's the best way to change the order of the core static library (`config.project.mk`) without messing up the other build systems?

**Details:**
Here's my hypothesis for what's messing up with the linking on the current oF make system on windows (with MSYS).  

**1)** Some libraries are missing (as compared to C::B build), so I tack them on to the end of the g++ call:

```
../../../libs/videoInput/lib/win_cb/videoInputLib.a ../../../libs/FreeImage/lib/win_cb/FreeImage.lib ../../../libs/quicktime/lib/win_cb/qtmlClient.lib ../../../libs/glew/lib/win_cb/glew32s.lib ../../../libs/glu/lib/win_cb/glu32.lib -lopengl32 -lgdi32 -lmsimg32 -lglu32 -ldsound -lwinmm -ldxguid -lstrmbase -lstrmiids -lz -luuid -lole32 -loleaut32 -lsetupapi -lwsock32 -lws2_32 -lIphlpapi -lComdlg32
```

These are easy to add in `config.win_cb.default.mk` using `PLATFORM_LIBRARIES` and `PLATFORM_STATIC_LIBRARIES`

**2)** I believe some of the libraries that are included are not in the proper order.  The order that I found that works (based on the C::B build commands):

```
libPocoNetSSL.a -> libPocoCrypto.a  -> libPocoNet.a  -> libPocoUtil.a  -> libPocoXML.a  -> libPocoFoundation.a  -> libPocoZip.a  -> libssl.a  -> libcrypto.a
```

I'm not sure how to fix this without modifying the common build process.  The order of these is set in `config.project.mk`.

**Comment by bakercp** [comment]
More great progress. I'll take a look this weekend.

**Comment by mikewesthad** [comment]
Didn't mean it to awake you from your ofSketch break.  I was bleary-eyed when I took a swing at this this morning, so I didn't poke around `config.win_cb.default.mk` before posting.  

I was able to get the Poco libs ordered.  The make system allows for ordering of libraries within a folder using libsorder.make.  Still, libssl and libcrypto needed to be fed into the linker after Poco.  The core libs are fed into the linker first, then the platform libs.  By adding ssl+crypto as platform libs in `config.win_cb.default.mk`, that fixed all the ordering issues.  (I'll fork your openFrameworks fork and submit a [PR](https://github.com/bakercp/openFrameworks/pull/4) with all this.)

Linking stage complete...but new errors encountered post-linkage.  I'll look into it later.  `install_name_tool: Command not found`

**Comment by mikewesthad** [comment]
Fixed + updated in the PR.  Build success :)

**Comment by bakercp** [comment]
!! :) :+1:

**Comment by mikewesthad** [comment]
Right, one more thing.  Make (via MSYS) is capable of multiple processes via the -jN flag, where N is the number of jobs.  The caveat is that the flag is broken, at least for me and [another](http://sourceforge.net/p/mingw/bugs/1950/), for recent versions of MSYS.  I rolled back to an earlier version as suggested in that link and the jobs works.  -j4 chopped the build time in half.

**Comment by bakercp** [comment]
How did you roll back to a previous version and what version did you use?

**Comment by mikewesthad** [comment]
I have two versions of MinGW installed on my machine, each in their own directory.  One is the most recent version and the other is the Code::Blocks version.  The most recent version has a get command, so I rolled back using `mingw-get upgrade msys-core-bin=1.0.17-1`.  Then I copy+pasted that MSYS folder into the Code::Blocks MinGW folder.

**Comment by mikewesthad** [comment]
It seems like the last item on the list, before we can label this build process as working, is to stop `make` from rebuilding all of core oF every time.  Throwing my hands up and asking the wider oF community ([thread with some details](http://forum.openframeworks.cc/t/windows-mingw-make-system-core-of-is-always-recompiled/16697)).

**Comment by bakercp** [comment]
@mikewesthad thanks for the epic work on this.  I know it is a very time-consuming process and it isn't the most gratifying work.  Believe me, I know!  If we can't get a good answer from the community, this might be a stackoverflow question ... :)

**Comment by mikewesthad** [comment]
Jeeze, what a simple fix it was for the above problem.  It's working now.  (small) [PR](https://github.com/bakercp/openFrameworks/pull/5) incoming.

@bakercp, no problem.  Happy to contribute.  Anything else that needs to be done in this issue to allow   ofSketch to be able to call the makefile and extract?  I've been using cmd to call make.  Setting up the environment variables for that isn't hard (and can be done in a `.bat`).  MinGW bin path can go into the user variable `Path`.  MSYS bin path needs to go in the _system_ variable `Path`, so that it can override the windows implementations of functions like find, kill, sort, etc.

**Comment by bakercp** [comment]
I'm going to try to get all of this running today!

**Comment by bakercp** [comment]
can you provide said .bat file? :)

---

http://christopherbaker.net

**Comment by bakercp** [comment]
by the way all of this great work is going to be super helpful for getting apothecary support for mingw / win_cb :)

**Comment by mikewesthad** [comment]
Exciting!  As for apothecary, that would be cool.  It was very unhappy with me when I tried to run it last week.  The commands I tried all errored out.

In writing a .bat, I realized the best thing would be executing the below command.  When called within a cmd, it will set the path correctly for MSYS+MinGW.  The changes are local only to that cmd (and any cmds spawned from it), so it would have to be called when ofSketch opens up the cmd that is doing the making.

```
SET PATH=C:\MinGW\msys\1.0\bin;%PATH%;C:\MinGW\bin
```

There are ways to do it permanently (using `setx PATH new_path_stuffs /M`), but when I tried them out cmd truncates the path according to a [1024 character limit](http://superuser.com/questions/387619/overcoming-the-1024-character-limit-with-setx).  Not a friendly way to handle an error....

**Comment by bakercp** [comment]
@mikewesthad so with `SET PATH=C:\MinGW\msys\1.0\bin;%PATH%;C:\MinGW\bin` -- will that set up the correct mingw g++ toolchain etc?  or do i need to set that up with an fstable somewhere?

**Comment by mikewesthad** [comment]
That should do it (or at least it worked on my machine).  I'm not familiar with fstable?

**Comment by bakercp** [comment]
OK.  Sorry, I think I need it broken down by steps -- I'm having a hard time piecing all of it together from the thread above.  Here's what I got so far:
- Install Codeblocks 12.11.
- Install [additional files](http://openframeworks.cc/setup/codeblocks/) for codeblocks.
- Copy `C:\Program Files (x86)\CodeBlocks\MinGW` to `C:\MinGW`
- Install The latest MSYS to `C:\MinGW_Latest`.
- Open `cmd.exe`, navigate to C:\MinGW_Latest\bin and `run mingw-get upgrade msys-core-bin=1.0.17-1`
- Copy the `C:\MinGW_Latest\msys` into `C:\MinGW`
- Close all running MSYS instances.
- Open `cmd.exe`.
  - Run `SET PATH=C:\MinGW\msys\1.0\bin;%PATH%;C:\MinGW\bin`
  - Change into the `openFrameworks/libs/openFrameworksCompiled/project` 
  - Run `make`

**Comment by mikewesthad** [comment]
The MinGW Code::Blocks has the [additional files](http://openframeworks.cc/setup/codeblocks/) from the oF installation process?

But yeah, that all looks right to me.  Is it erroring out at some point?

**Comment by bakercp** [comment]
No errors yet, I've only written down the instructions and I'm trying to figure out how to get administrator rights do delete directories on my windows machine ... fun times.  Does the command need to be done like this? 
`
SET PATH=C:\MinGW\msys\1.0\bin;%PATH%;C:\MinGW\bin; make
`

I'm not sure of the syntax for such things on windows.

**Comment by bakercp** [comment]
(also ignore the thing about fstable)

**Comment by mikewesthad** [comment]
Ah okay.  Well I'm following along on a fresh install of windows, trying to bypass needing Code::Blocks.  

I think that if you want to combine the two commands on windows, cmd uses `&` instead of `;`.  But you don't need to throw them on the same line.  The PATH is a variable that is tied to the instance of  cmd.

**Comment by bakercp** [comment]
Nice.  If you can figure out how to do it by ming-getting the older equivalent codeblocks 12.11 mingw setup that would be awesome.

**Comment by mikewesthad** [comment]
Yeah, I'll see if I can get it working.

I have a possible correction for step 5: _Open MSYS from C:\MinGW_Latest and and `run mingw-get upgrade msys-core-bin=1.0.17-1`_
I think it should be: _Open CMD, navigate to C:\MinGW_Latest\bin and `run mingw-get upgrade msys-core-bin=1.0.17-1`_

**Comment by mikewesthad** [comment]
Sigh, mild stumbling block.  The path command (`SET PATH=C:\MinGW\msys\1.0\bin;%PATH%;C:\MinGW\bin`) works find on my Windows 7 machine.  Just retested it.  But it doesn't work on my new machine, Windows 8.1.  On Windows 7, prefixing %PATH% with the msys bin allows the win implementation of `find` to be overriden with the msys version.  On Windows 8.1, security must have been tightened up.  Looking for a solution now...

**Comment by bakercp** [comment]
speaking of 8.1 ... how do I delete files on windows?

![windows_8_-_parallels_desktop](https://cloud.githubusercontent.com/assets/300484/3885440/170a69ce-21c2-11e4-8890-32e8b1bf8d51.png)

I've tried just about everything and it won't let me delete my old install ...

**Comment by bakercp** [comment]
I am WINDOZE\bakercp ... why do i need my own permission?  And how would I give it anyway?

**Comment by bakercp** [comment]
OK.  I'm going to go take a bike ride.  I've spent over an hour trying to delete files on windows.  This is friggin' insane.  I'll be back on in a bit.

**Comment by mikewesthad** [comment]
Right click > properties > security.  That lists all the user permissions for the folder.  Is deleting not enabled for the WINDOZE\backercp user?

Understood. :( silly windows.

**Comment by bakercp** [comment]
![windows_8_-_parallels_desktop](https://cloud.githubusercontent.com/assets/300484/3885499/6b90dfe0-21c3-11e4-99fb-3ac25a7490e6.png)

**Comment by mikewesthad** [comment]
...I'm stumped without being able to see the machine.  I guess you already tried a reboot?

**Comment by bakercp** [comment]
Yeah, I reboot.  Regular scheduled reboots are part of my standard windows workflow.  Doesn't every windows user do that?

**Comment by mikewesthad** [comment]
My reboot count is uncountable.  I guess the next step is to open up your hard drive and manually flip the bits.

**Comment by mikewesthad** [comment]
I'm calling it quits on this Win 8 stuff for the night.
1. CMD route: I can't get cmd to have the right path information to override `find`.
2. Powershell route: `$env:Path = "C:\MinGW\msys\1.0\bin;" + $env:Path + ";C:\MinGW\bin"` sets the path, but the makefile is throwing errors.  `$(MAKE)` is set to some wonky command.  I can't find where `MAKE` is defined.  Do you have a clue?  
3. MSYS route: didn't try diving back into that

**Comment by bakercp** [comment]
The internet looped back on itself just now:

http://stackoverflow.com/questions/25255504/set-system-path-for-cmd-on-windows-8-1-trying-to-override-windows-system32-find/25256551#25256551

Also, I successfully deleted my files by manually deleting each one.  Fun times.  A long night bikeride helped.

**Comment by mikewesthad** [comment]
That sounds awful.  Could it be boot camp/parallels?  The time I've had that sort of error is usually from some hidden process that froze while using the folder.  (I'm looking at you Dropbox.)

And I am quite thankful that you solved my silly CMD issue.  I was going a bit insane.  But now building on Windows 8.1 works **both** with the C::B MinGW and with the latest version of MinGW.  No need to use C::B methinks.  (So long as the proper additions are copied over from the C::B oF installation page.)  Presumably, the same applies to Windoze 7, though I only tested with the C::B MinGW.

 I did have to make a couple changes to get it to work on win8:  
- Added another `else ifeq` to config.shared.mk to detect Windows 8 (i.e. MINGW32_NT-6.2)
- Removed libz and strmbase from `PLATFORM_LIBRARIES` in config.win_cb.default.mk.  I don't know if those are necessary, since I just copied them over form C::B build instructions.  Still, I'm not sure why they worked on my Windows 7 box.  My best guess is that it has to do with the linker search paths (`ld --verbose | grep SEARCH_DIR`).  They aren't set up to point to the MinGW directories on my Windows 8 machine.  And that's at least where libz is.  That's one of the C::B additions from the oF installation.

**Comment by mikewesthad** [comment]
Oh and I see that you were using MSYS2.  Should I switch over to that?

**Comment by bakercp** [comment]
@arturoc we're making progress on this -- can you explain where the codeblocks-dependencies came from?  Is that something you compiled at some point?  Is it something that we should add to apothecary?

@mikewesthad Not sure about MSYS2 yet -- it's pretty nice.  I'm doing my testing with it today and will see if it works easily.  The main advantage is that it has a richer package management system (i.e. you can get curl, etc) and supports 64bit compilation, which we'll want to do eventually.  Still trying to get my brain around it all ...

**Comment by bakercp** [comment]
![windows_8](https://cloud.githubusercontent.com/assets/300484/3893675/253b175c-223d-11e4-9d72-c3eb43f39ce4.png)

A truism.  Jenny Holzer would be proud.

**Comment by mikewesthad** [comment]
I don't even...

Also, I moved the instructions for getting a MinGW+MSYS build system working over to a [public hackpad](https://hackpad.com/MinGW-MSYS-openFrameworks-Setup-GuhEod9okeN).  Probably makes more sense as a .md somewhere in the repo.

**Comment by bakercp** [comment]
@mikewesthad @arturoc are the additional dependencies just binaries that can be installed with the mingw installer?  Just trying to understand the full process ...

**Comment by mikewesthad** [comment]
Good question for @arturoc.  I only recognized a name or two in the additional Code::Block dependencies.  I saw some Direct3D extension libraries and a compression library.

**Comment by bakercp** [comment]
OK I followed the hackpad instructions to the letter and got everything compiled, but had some linking failures when i actually got around to linking the app:
![windows_8](https://cloud.githubusercontent.com/assets/300484/3897120/11d3a8d8-225f-11e4-9891-08dd35976313.png)

Seen anything like that?  Might this be a mingw issue related to not using the CB 12.11 MingW version?

**Comment by mikewesthad** [comment]
Hmm, I have seen errors like that.  Maybe the order of libraries in the linker call needs to be changed.  Latest MinGW version built for me, so let me do a fresh clone of openFrameworks and investigate.

**Comment by bakercp** [comment]
Cool.  All of your latest updates should be in the master openframeworks repo.

**Comment by bakercp** [comment]
ofSketch was just successfully compiled using mingw and makefiles on windows :)  This means ofSketch will be able to use the same system to compile sketches on windows!

A bit of work needs to be done to get ride of the *nix / posix dependencies, but we are well on our way!

**Comment by mikewesthad** [comment]
Yessssss! 

I'm curious to hear about (or be pointed towards the relevant code) with the posix dependencies

**Comment by bakercp** [comment]
@brannondorsey is currently working on implementing Poco::Path for crossplatform paths.

I am working on a way to do a non-blocking `std::getline` on windows without ::select or ::poll 

https://github.com/olab-io/ofSketch/blob/develop/ofSketchApp/src/BaseProcessTask.cpp#L73-L85

If you've got any ideas ... 

These data structures are available in `<winsock.h>`, but usually that is used with SOCKETs rather than the pipe handle ...

**Comment by mikewesthad** [comment]
Mmm, nothing helpful coming to mind at the moment, but I'll take a closer look at the code this week.

**Comment by bakercp** [comment]
OK.  I have an MinGW folder that works and I successfully place the folder in an arbitrary location on another computer and it seems to be working.  Eventually it will go in the app bundle on the windows distro.  For now, it won't live in the ofSketch repo, but, like the openFrameworks distros, we'll download it, unzip it and package it with each release.

I'll upload it somewhere soon and get it integrated into the build ofSketch build scripts -- @mikewesthad if you want it for testing, let me know and I'll make sure it's up somewhere ASAP.

**Comment by mikewesthad** [comment]
@bakercp that all sounds fantastic.  I have to work on some other projects for a few days, but I would like to help in some capacity after that

**Comment by bakercp** [comment]
Who hoo!  It's here.

https://vimeo.com/106986055

Now just have to merge it all in and take care of a few loose ends.  Thanks for all the help on this.

**Comment by mikewesthad** [comment]
:) :) :) :) :)

Can't wait to try this.

**Comment by bakercp** [comment]
Added.  And closed!
